File: src/com/suushiemaniac/cubing/alglib/move/ClockMove.java
package com.suushiemaniac.cubing.alglib.move;

import com.suushiemaniac.cubing.alglib.move.modifier.ClockDirectionModifier;
import com.suushiemaniac.cubing.alglib.move.modifier.ClockModifier;
import com.suushiemaniac.cubing.alglib.move.modifier.ClockNumModifier;
import com.suushiemaniac.cubing.alglib.move.plane.ClockPlane;
import com.suushiemaniac.cubing.alglib.transform.Transform;

public class ClockMove implements Move {
    private final ClockPlane plane;
    private final ClockModifier modifier;
    private final boolean endPinConfig;

    public ClockMove(ClockPlane plane) {
        this.plane = plane;
        this.modifier = ClockModifier.endPinModifier;
        this.endPinConfig = true;
    }

    public ClockMove(ClockPlane plane, ClockNumModifier numModifier, ClockDirectionModifier directionModifier) {
        this.plane = plane;
        this.modifier = new ClockModifier(numModifier, directionModifier);
        this.endPinConfig = false;
    }

    public ClockMove(ClockPlane plane, ClockModifier modifier) {
        this.plane = plane;
        this.modifier = modifier;
        this.endPinConfig = false;
    }

    @Override
    public boolean merges(Move other) {
        return other instanceof ClockMove
                && this.mayAppend(other)
                && !((ClockMove) other).isEndPinConfig()
                && this.plane == other.getPlane()
                && !this.cancels(other);
    }

    @Override
    public boolean cancels(Move other) {
        return other instanceof ClockMove
                && this.mayAppend(other)
                && !((ClockMove) other).isEndPinConfig()
                && this.plane == other.getPlane()
                && this.modifier.toNumber() + ((ClockMove) other).getModifier().toNumber() == 0;
    }

    @Override
    public boolean mayAppend(Move other) {
        return !this.endPinConfig;
    }

    @Override
    public ClockPlane getPlane() {
        return this.plane;
    }

    @Override
    public ClockModifier getModifier() {
        return this.modifier;
    }

    @Override
    public ClockMove merge(Move other) {
        if (!(other instanceof ClockMove) || this.endPinConfig) return null;
        return new ClockMove(this.plane, this.modifier.merge(other.getModifier()));
    }

    @Override
    public int getDepth() {
        return 1;
    }

    @Override
    public Move inverse() {
        return this.endPinConfig ? new ClockMove(this.plane) : new ClockMove(this.plane, this.modifier.inverse());
    }

    @Override
    public Move transform(Transform transform) {
        return transform.transform(this);
    }

    @Override
    public String toString() {
        return this.toFormatString();
    }

	@Override
	public boolean equals(Object obj) {
		return obj instanceof ClockMove
                && ((ClockMove) obj).modifier.equals(this.modifier)
                && ((ClockMove) obj).plane == this.plane
                && ((ClockMove) obj).endPinConfig == this.endPinConfig;
	}

	@Override
	public int hashCode() {
		return this.toFormatString().hashCode();
	}

	@Override
    public String toFormatString() {
        if (this.endPinConfig) return this.plane.toFormatString();
        else return this.plane.toFormatString() + (this.plane.isRotation() ? this.modifier.getNumModifier().toFormatString() : this.modifier.toFormatString());
    }

    public boolean isEndPinConfig() {
        return this.endPinConfig;
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/move/ClockMove.kt
package com.suushiemaniac.cubing.alglib.move

import com.suushiemaniac.cubing.alglib.move.modifier.ClockDirectionModifier
import com.suushiemaniac.cubing.alglib.move.modifier.ClockModifier
import com.suushiemaniac.cubing.alglib.move.modifier.ClockNumModifier
import com.suushiemaniac.cubing.alglib.move.plane.ClockPlane

data class ClockMove(override val plane: ClockPlane, override val modifier: ClockModifier, val isEndPinConfig: Boolean) : Move {
    override val notation = if (this.isEndPinConfig) this.plane.notation else this.plane.notation + if (this.plane.isRotation) this.modifier.numModifier.notation else this.modifier.notation

    override val depth = 1

    constructor(plane: ClockPlane) : this(plane, ClockModifier.endPinModifier, true)
    constructor(plane: ClockPlane, numModifier: ClockNumModifier, directionModifier: ClockDirectionModifier) : this(plane, ClockModifier(numModifier, directionModifier), false)
    constructor(plane: ClockPlane, modifier: ClockModifier) : this(plane, modifier, false)

    override fun merges(other: Move): Boolean {
        return (other is ClockMove
                && !this.isEndPinConfig
                && !other.isEndPinConfig
                && this.plane == other.plane
                && !this.cancels(other))
    }

    override fun cancels(other: Move): Boolean {
        return (other is ClockMove
                && !this.isEndPinConfig
                && !other.isEndPinConfig
                && this.plane == other.plane
                && this.modifier.toNumber() + other.modifier.toNumber() == 0)
    }

    override fun mayAppend(other: Move) = !this.isEndPinConfig
    override fun merge(other: Move) = ClockMove(this.plane, this.modifier.merge(other.modifier))
    override fun inverse() = if (this.isEndPinConfig) ClockMove(this.plane) else ClockMove(this.plane, this.modifier.inverse())

    override fun toString() = this.notation
}